feat: DAO SDK

n0izn0iz opened the pull request:
This implement a DAO SDK inspired by [DA0-DA0](https://github.com/DA0-DA0/dao-contracts)

The code in this PR has been extracted from https://github.com/gnolang/gno/pull/1154 and cleaned (reviews on the previous PR should have been addressed)

It includes:
- Interfaces for the DAO `gno.land/p/demo/dao_maker/dao_interfaces`
- A reference implementation for the core of the DAO `gno.land/p/demo/dao_maker/dao_core`
- A membership-based voting module `gno.land/p/demo/dao_maker/dao_voting_group`
- A single-choice proposal module `gno.land/p/demo/dao_maker/dao_proposal_single`
- An example DAO `gno.land/r/demo/dao_maker/dao_realm`
- A registry to dynamically interact with any registered DAO using it's address `gno.land/r/demo/dao_maker/dao_registry`

For usage see the [tutorial](https://github.com/TERITORI/gno/blob/1555f35be0b9ad8f65c2179d9a9b8c8f2a79aff4/docs/how-to-guides/creating-dao.md), it is integrated into docusaurus

<img width="1505" alt="Screenshot 2024-04-12 at 18 41 10" src="https://github.com/gnolang/gno/assets/7917064/403a27c4-24b1-493f-af89-f35591c1aece">

---

There is an example membership-based DAO at `gno.land/r/demo/teritori/dao_realm` that can administer itself and a grc20 token at `gno.land/r/demo/teritori/tori`

Here is it's GnoWeb Render:
<img width="1219" alt="Screenshot 2024-04-12 at 18 28 41" src="https://github.com/gnolang/gno/assets/7917064/16c19c21-5d9e-426f-82ed-7252c7e7cd06">

---

This powers the "Organizations" feature of Teritori dApp on Gno as shown below

The DAO creation UI uses a Gno template and instantiates a new realm.

<img width="1264" alt="Screenshot 2024-04-12 at 18 43 23" src="https://github.com/gnolang/gno/assets/7917064/c5c412bf-a26c-4af7-9196-03c259241125">
<img width="1511" alt="Screenshot 2024-04-12 at 18 44 50" src="https://github.com/gnolang/gno/assets/7917064/0c69d902-b7fd-46b0-a294-93108655c01e">
<img width="1512" alt="Screenshot 2024-04-12 at 18 44 15" src="https://github.com/gnolang/gno/assets/7917064/077fea68-e76f-4cea-8542-fefcfdf215b9">
<img width="1507" alt="Screenshot 2024-04-12 at 18 42 26" src="https://github.com/gnolang/gno/assets/7917064/4535b351-816d-4afb-8691-60ead6572b85">

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

notJoon commented on examples/gno.land/p/demo/teritori/dao_core/dao_core.gno:
```diff
@@ -0,0 +1,188 @@
+	return d.activeProposalModuleCount
+}
+
+func (d *daoCore) Render(path string) string {
+	s := "# DAO Core\n"
+	votingInfo := d.votingModule.Info()
+	s += "## Voting Module: " + votingInfo.String() + "\n"
+	s += d.votingModule.Render("")
+	s += "## Supported Messages:\n"
+	s += d.registry.Render()
+	s += "## Proposal Modules:\n"
+	for i, propMod := range d.proposalModules {
+		if !propMod.Enabled {
+			continue
+		}
+		info := propMod.Module.Info()
+		s += "### #" + strconv.Itoa(i) + ": " + info.String() + "\n"
+		s += propMod.Module.Render("")
+	}
+	return s + "\n"
+}
```
When concatenating strings, using `strings.Builder` instead of the `+=` operator can reduce unnecessary memory usage.

```suggestion
func (d *daoCore) Render(path string) string {
    var sb strings.Builder
    sb.WriteString("# DAO Core\n")

    votingInfo := d.votingModule.Info()
    sb.WriteString("## Voting Module: ")
    sb.WriteString(votingInfo.String())
    sb.WriteString("\n")

    sb.WriteString(d.votingModule.Render(""))

    sb.WriteString("## Supported Messages:\n")
    sb.WriteString(d.registry.Render())

    sb.WriteString("## Proposal Modules:\n")
    for i, propMod := range d.proposalModules {
        if !propMod.Enabled {
            continue
        }
        info := propMod.Module.Info()
        sb.WriteString("### #")
        sb.WriteString(strconv.Itoa(i))
        sb.WriteString(": ")
        sb.WriteString(info.String())
        sb.WriteString("\n")
        sb.WriteString(propMod.Module.Render(""))
    }

    sb.WriteString("\n")
    return sb.String()
}
```

notJoon commented on examples/gno.land/p/demo/teritori/dao_interfaces/messages_registry.gno:
```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+func (r *MessagesRegistry) Render() string {
+	s := ""
+	r.handlers.Iterate("", "", func(key string, value interface{}) bool {
+		s += "- " + value.(MessageHandler).Type() + "\n"
+		return false
+	})
+	return s
+}
```
For the same reason, it would be good to use `strings.Builder` here as well.
```suggestion
func (r *MessagesRegistry) Render() string {
    var sb strings.Builder

    r.handlers.Iterate("", "", func(key string, value interface{}) bool {
        sb.WriteString("- ")
        sb.WriteString(value.(MessageHandler).Type())
        sb.WriteString("\n")
        return false
    })

    return sb.String()
}
```

notJoon commented on examples/gno.land/p/demo/teritori/dao_proposal_single/dao_proposal_single.gno:
```diff
@@ -0,0 +1,410 @@
+	switch d.opts.Threshold.(type) {
+	case *ThresholdThresholdQuorum:
+		threshold := d.opts.Threshold.(*ThresholdThresholdQuorum)
+		thresholdStr = "Threshold: " + threshold.Threshold.String() + "\n\n" +
+			"Quorum: " + threshold.Quorum.String()
```
It seems we can use `ufmt.Sprintf` for formatting in this part.

```suggestion
		thresholdStr = ufmt.Sprintf("Threshold: %s\n\nQuorum: %s\n\n", threshold.Threshold.String(), threshold.Quorum.String())
```

notJoon commented on docs/how-to-guides/creating-dao.md:
````diff
@@ -0,0 +1,333 @@
+```go
+package my_dao
+
+import (
+    "gno.land/p/demo/teritori/dao_interfaces"
+    "gno.land/p/demo/teritori/dao_voting_group" // <- new
+)
+
+func init() {
+    var group *dao_voting_group.VotingGroup // <- new
+
+    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
+        group = dao_voting_group.NewVotingGroup() // <- new
+    }
+}
````
It's not crucial, but what do you think about marking changes in the code snippet like diff? 

For example, line that are added or modifed would be marked with a `+` sign, and lines that are deleted would be marked with a `-` sign. The downside in this case would be that we cannot use syntax highlighting.

Example:

```diff
import (
    "gno.land/p/demo/teritori/dao_interfaces"
+    "gno.land/p/demo/teritori/dao_voting_group"
)

func init() {
+    var group *dao_voting_group.VotingGroup

    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
+        group = dao_voting_group.NewVotingGroup()
    }
}
```

n0izn0iz replied:
I prefer to keep syntax highlighting and pastability

leohhhn reviewed (commented):
Thank you for your awesome contribution, and thank you for helping make the docs better! I left reviews on the tutorial, mainly about the flow of it and typos. I haven't finished the review fully - I still want to go through the full tutorial hands-on. I am submitting this to save progress on the review :)

n0izn0iz commented:
Thanks for the review @leohhhn I'm addressing your comments

I have two remarks to spark discussion around guidelines:

- Adding periods everywhere is [considered a bad ux practice by google](https://m3.material.io/foundations/content-design/style-guide/grammar-and-punctuation#555b17c7-de48-4974-8a82-51106899dc0e)

- Tutorials should be playful instead of formal
  > The use of humor in the college classroom has been researched extensively (see Segrist & Hupp, 2015 who summarized 41 years of literature on humor in the college classroom) and has been shown to have many benefits for students (Banas, Dunbar, Rodriguez & Liu, 2011; Garner, 2006; Huss & Estep, 2016; Pollio, 2002), some of which are:
    An increase in learning.
    An increase in self-motivation.
    An increase in class attendance.
    An increase in test performance.
    An increase in divergent thinking.
    An increase of interest in learning.
    A reduction of anxiety and stress in dealing with difficult material.
    The creation of a positive social and emotional learning environment.
    The creation of a common psychological bond between students and faculty.

elix1er commented:
impressive work, keep grinding

leohhhn commented on docs/how-to-guides/creating-dao.md:
```diff
@@ -0,0 +1,343 @@
+
+## Overview
+
+This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the Realm, please see the
```
```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

leohhhn replied:
```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

leohhhn commented on docs/how-to-guides/creating-dao.md:
```diff
@@ -0,0 +1,343 @@
+
+### Message handlers
+
+Proposals actions are encoded as objects implementing `gno.land/p/demo/dao_maker/dao_interfaces.ExecutableMessage`.
```
```suggestion
Proposal actions are encoded as objects implementing `ExecutableMessage` found under `gno.land/p/demo/dao_maker/dao_interfaces`.
```

leohhhn commented on docs/how-to-guides/creating-dao.md:
````diff
@@ -0,0 +1,343 @@
+}
+```
+
+They are deserialized and executed by message handlers implementing `gno.land/p/demo/dao_maker/dao_interfaces.MessageHandler`.
````
```suggestion
They are unmarshalled and executed by message handlers implementing `gno.land/p/demo/dao_maker/dao_interfaces.MessageHandler`.
```
Marshal/unmarshal is a more "go way" to say serialize/deserialize.

leohhhn commented on docs/how-to-guides/creating-dao.md:
```diff
@@ -0,0 +1,343 @@
+Message handlers are registered at core creation and new message handlers can be registered via proposals to extend the DAO capabilities.
+
+## Practical Implementation
+
```
```suggestion
In this section, we will showcase how to implement your own DAO using the `dao_maker` package suite.

```

leohhhn commented on docs/how-to-guides/creating-dao.md:
```diff
@@ -0,0 +1,343 @@
+#### Create a new Gno module
+
+- Create a new directory and move into it: `mkdir my-gno-dao && cd my-gno-dao`
+- Initialize the gno module: `gno mod init gno.land/r/demo/my_dao`
```
```suggestion
- Initialize the gno module: `gno mod init gno.land/r/<your_namespace>/my_dao`
```

leohhhn reviewed (approved):
Left some minor suggestions to the docs page. Otherwise we should be good to go.

Sorry for the delay, and great work.

moul commented on examples/gno.land/p/demo/users/types.gno:
```diff
@@ -12,3 +16,14 @@ func (aon AddressOrName) GetName() (string, bool) {
 	return "", false
 }
+
+func (aon AddressOrName) IsValid() bool {
```
what about using `users.Resolve(aon).IsValid()`?

n0izn0iz replied:
can't call the realm method `users.Resolve` from a `p` package

moul commented on examples/gno.land/p/demo/dao_maker/havl/havl.gno:
```diff
@@ -0,0 +1,133 @@
+package havl
```
can you add more info about what is this package about?

it looks like something that should be located in `p/demo/havl`.

+ add some tests.

moul commented on examples/gno.land/p/demo/dao_maker/jsonutil/gno.mod:
```diff
@@ -0,0 +1,7 @@
+module gno.land/p/demo/dao_maker/jsonutil
```
The current `p/demo/json` library should be suitable for your needs.

I'm against having any JSON library written in userland because I expect that we can use @gfanton's #1776 and avoid any boilerplate code that isn't either for humans (rendering) or for other contracts (imports). Generally, anything about marshalling doesn't fall into these categories.

In the meantime, you can reuse the existing JSON library if needed. However, as soon as possible, I expect we can delete it completely. Additionally, if possible, I expect we don't merge any new JSON library in the meantime.

n0izn0iz replied:
this extends `p/demo/json`

moul commented on examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:
```diff
@@ -0,0 +1,173 @@
+func (v *VotingGroup) Render(path string) string {
+	sb := strings.Builder{}
+	sb.WriteString("Member count: ")
+	sb.WriteString(strconv.FormatUint(uint64(v.MemberCount(havl.Latest)), 10))
```
```suggestion
	sb.WriteString(ufmt.Sprintf("%d", havl.Latest))
```

moul commented on examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:
```diff
@@ -0,0 +1,173 @@
+
+func (g *VotingGroup) GetMembers(start, end string, limit uint64, height int64) []Member {
+	var members []Member
+	g.powerByAddr.Iterate(start, end, height, func(k string, v interface{}) bool {
+		if limit > 0 && uint64(len(members)) >= limit {
+			return true
+		}
```
You can use `avl.Tree.IterateByOffset`.

moul commented on examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:
```diff
@@ -0,0 +1,173 @@
+}
+
+func (g *VotingGroup) MemberCount(height int64) uint32 {
+	val, ok := g.memberCount.Get("", height)
```
isn't `avl.Tree.Size` enough for you?

moul commented on examples/gno.land/r/demo/dao_maker/tori/tori.gno:
can you move this to `gno.land/r/teritori/tori`?

moul commented on examples/gno.land/p/demo/dao_maker/dao_interfaces/messages.gno:
```diff
@@ -0,0 +1,21 @@
+	"gno.land/p/demo/json"
+)
+
+type ExecutableMessage interface {
+	ToJSON() *json.Node
+	FromJSON(ast *json.Node)
+
+	String() string
+	Type() string
+}
```
why is this needed?

moul reviewed (changes requested):
There are several interesting aspects in that PR, but there is too much work to review at the moment. Therefore, I will mark it as "request changes" to allow more time for a thorough review.

A high-level suggestion I can provide is to:
- Simplify the codebase by reducing the number of packages, files, functions, types, and boilerplate, and avoiding the use of JSON, etc. Consider merging some packages together.
- Improve the naming conventions. For instance, use more specific names like "Teritori Realms" and "Daomaker" (a better name could facilitate faster merging, as "Daomaker" is currently too generic and sounds overly official, while "Teritori" would make it easier to integrate).

moul commented:
Can you move all the packages under `{r,p}/teritori/...` and relocate the `docs/...` into `r/teritori/dao_maker/README.md`?

I believe we should merge your PR in its current state but under teritori.

Gno2D2 commented:
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🔴 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))
 🔴 Changes to 'docs' folder must be reviewed/authored by at least one devrel and one tech-staff


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (TERITORI:daosdk) is up to date with base (master): behind by 490 / ahead by 111

```
</blockquote></details>

<details><summary>Changes to 'docs' folder must be reviewed/authored by at least one devrel and one tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 A changed file matches this pattern: ^docs/ (filename: docs/how-to-guides/creating-dao.md)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Or
    ├── 🔴 And
    │   ├── 🔴 Pull request author is a member of the team: devrels
    │   └── 🔴 At least 1 user(s) of the team tech-staff approved pull request
    └── 🔴 And
        ├── 🔴 Pull request author is a member of the team: tech-staff
        └── 🟢 At least 1 user(s) of the team devrels approved pull request

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

n0izn0iz commented:
closing in favor of #3358 
will address reviews there
